internal/jimm/juju/model_upgrader.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2025 Canonical.
+
+package juju
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/db"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/servermon"
+)
+
+// TickModelUpgrades loops over models due for upgrade, checks their status
+// and runs the logic for the next step in the upgrade process.
+// This effectively implements progressing the state machine for each model's upgrade.
+func (j *JujuManager) TickModelUpgrades(ctx context.Context) error {
+	const op = "jimm.juju.TickModelUpgrades"
+	zapctx.Info(ctx, string(op))
+	durationObserver := servermon.DurationObserver(servermon.JimmMethodsDurationHistogram, string(op))
+	defer durationObserver()
+
+	// Obtain likely model upgrade IDs in a separate read transaction
+	upgrades, err := j.Database.GetAllModelUpgrades(ctx)
+	if err != nil {
+		return errors.E(op, err)
+	}
+
+	for _, upgrade := range upgrades {
+		err := j.Database.Transaction(func(database *db.Database) error {
+			if err := database.GetModelUpgradeForUpdate(ctx, &upgrade); err != nil {
+				// Ignore missing entries, they may have been concurrently deleted
+				if errors.ErrorCode(err) != errors.CodeNotFound {
+					return errors.E(op, err)
+				}
+			}
+
```

Comment:
Here we'd need to store/fetch a retry count or something similar to decide when to fail.